### PR TITLE
Better Reblogs — reblogs layout extension

### PR DESCRIPTION
**Pull request by nightpool**
Better Reblogs merges better_reblog_layout by @macleodsawyer, old_reblogs by @ewenig, and reblog tweaks by myself.

I also added a few more options and features not found in either, and made sure most of the options played harmoniously together layout-wise.

### CODE REVIEW

**Comment by sheridan101** [comment]
is this anything like http://thetimeisneveright.tumblr.com/post/128224247395/here-is-the-unofficial-xkit-extension-to-fix because that seems to be working pretty good for me

**Comment by ThePsionic** [comment]
Yes, once this is live you can select the "nested" style which restores the "old" blockquote look.

**Comment by Wolvan** [comment]
Tested and looks good to me! Nice work, @nightpool Really well done
@homu r+

**Comment by Wolvan** [comment]
@homu r-

**Comment by Wolvan** [comment]
Phew, was able to prevent the merge. Anyways, I think I found a bug, changing settings from flat to nested and back dupes the description, I'll look into that and give more information

**Comment by nightpool** [comment]
Yeah my destroy code isn't great I'll push something better.

**Comment by Wolvan** [comment]
Good to know, as I said it looks really good to me otherwise

**Comment by nightpool** [comment]
Hey @macleodsawyer and @ewenig, interested in your input on this.

**Comment by ewenig** [comment]
As for the nested code:

1. It doesn't handle reblog chains with a deactivated user exactly correctly. It shows up as a link with text `exampleblogDeactivated`, because `Deactivated` is part of the value returned from the `.text()` element of `$('.reblog-tumblelog-name')` at line 204. You'll have to implement something like I did here: https://github.com/ewenig/XKit/blob/old_reblogs/Extensions/old_reblogs.js#L77 Also you probably want to not throw a `href` element in the `a` tag when it's associated to a deactivated user, but that's just a nitpick
2. It does work in the inline blog viewer (what did you sacrifice to the DOM gods to get that working?) but clobbers all the content of text posts, resulting in this unfortunate mess:  No clue why it's doing this, I can do a little investigating today.
3. The nested HTML inserted in lines 221 and 223 has to be removed by the destructor. (It might just be cleaner to destroy and re-init the extension when the user moves from `flat` to `nested` and vice versa.)

**Comment by nightpool** [comment]
@ewenig in backwards order:

3) yep, @Wolvan mentioned this and will do. xkit_preferences already does destroy and reinits the extension.

2) do you have the iframe-peepr or the non-iframe peepr? Can you link a blog where this happens? I can't reproduce.

1) will fix

**Comment by ewenig** [comment]
@nightpool I have the non-iframe peepr, and it happens to every blog I open in the inline viewer.

**Comment by macleodsawyer** [comment]
Sorry, just got online.

Have any of you figured out a better destroy for the switching? I have been messing with it, and all I can think of is creating it's own JQuery redraw on the reblogs and that slightly works better.

**Comment by nightpool** [comment]
Hmm? The only thing I need to undo for the switching is delete the
old_reblogs div—everything else is handled by xkit. (removing the css that
hides the old posts, for example)

**Comment by macleodsawyer** [comment]
Wow, I should really finish my coffee before I try to do anything productive.

That's a way better fix.

**Comment by janitamontani** [comment]
I don't know how to install this

**Comment by bvtsang** [comment]
@janitamontani In order to test this, what you can do is:

1. View the [raw code](https://raw.githubusercontent.com/nightpool/XKit/reblog_tweaks/Extensions/better_reblogs.js) for Better Reblogs (I've linked you directly for convenience, but you can also access it via steps below)
  1. View "Files changed" link near the top of this page
  2. Click on the View button for `Extensions/better_reblogs.js`
  3. Click on the Raw button
2. On Tumblr, click the XKit button to bring up the control panel
3. Near the bottom, click on the Other tab > XKit Editor > Open Editor
4. On the upper left corner, click New Extension
5. Name the extension `better_reblogs`
6. Copy and paste the contents of the raw code from step 1 into the area where the autogenerated code is
7. On the left side of the screen, click Save
8. Go back to your Tumblr dashboard (or refresh it if it's on a separate tab)
9. Click the XKit button to bring up the control panel
10. Scroll down to Better Reblogs to configure your options

Remember, what you're testing is _not stable and may be buggy_. If you do find an issue, please don't hesitate to report it here so that we can fix it.

**Comment by ThePsionic** [comment]
A few more suggestions for flat mode, if you don't mind:
* An option to make the line between reblogs either thicker or otherwise more prominent
* An option to make the first post (OP) in the reblog tree more prominent than the rest of the comments

**Comment by janitamontani** [comment]
It didn't work. The dash looks back to normal but I still can't erase the other reblogs.

**Comment by Meektepp** [comment]
Chiming in to agree with ThePsionic on making the first post [OP] more prominent in the flat mode option. I've read every tumblr post tagged "tumblr update" during the last three days, and the ability to distinguish the original poster's text has been much mentioned, coming in second only after "remove avatars." The question is how.  In the visual mockup I saw, flat view indented reblog comments slightly (and equally), but to my old tired eyes that was not quite enough. All I can think of is a light grey background, and/or making the line between reblogs thicker after the OP.

**Comment by coderpatsy** [comment]
@janitamontani This extension is afaik only meant to change the look of reblogged posts on your dashboard. If you want to actually edit reblogs, try the already-released Editable Reblogs extension.

**Comment by anakhasilver** [comment]
I'm probably the person who asked for darker and thicker blockquotes. Personally, both me and my fiancee have trouble with both thin and light grey lines. They hurt our heads, so yes, something darker and a few pixels thicker would be appreciated as an option!

**Comment by rimbaum** [comment]
Just a suggestion, to avoid confusion by users at large. Perhaps rename the extension from "Better Reblogs" to something else? People will confuse this extension for Editable Reblogs, unless we add the ability to edit reblogs as an option within this extension instead of keeping Editable Reblogs as it is.

**Comment by Wolvan** [comment]
@rimbaum's got a point. I am not really happy with the current title either. I'd keep ER and this split apart though

**Comment by rimbaum** [comment]
I know there was talk of restructuring Tweaks, so since there's several 'classic post' type options in Tweaks, maybe remove those, add them into this one, and rename the extension to "Classic Post Styles" ?

**Comment by Zarggg** [comment]
Perhaps "Inline Reblogs" or "Threaded Reblogs" would be a better name?

**Comment by coderpatsy** [comment]
I like "Classic Reblog Styles".

**Comment by ThePsionic** [comment]
Classic Comments maybe?

**Comment by Wolvan** [comment]
I like the sound of classic reblog styles

**Comment by bvtsang** [comment]
From an ask, and seems like something to be aware of:

> Okay. With the better reblogs extension for the nested look for the dash on some posts that were made because of the new layout there is a long block of just lines leading to the comments. Which I will take over the stupid layout anyday, but when you get to the comments it doesn't show all the comments that were made. It cuts off a good portion of them, like they extend past the white of the post so we can't see it.

**Comment by amandorfer** [comment]
I just installed the better_reblogs extension manually and I'm messing with the options within the extension. It seems that when I choose "nested" the reblogs/words inside the post disappear. The post is still there, and it's there when i load the page, but once X-kit kicks in the content disappears. Just wanted to give you a heads up

(I figured it was better to pot here than to send an ask)

**Comment by macleodsawyer** [comment]
"Classic Reblog" has my vote since it fits with the other "Classic" extensions.

**Comment by nightpool** [comment]
I considered classic reblog, but this styles more things then just a
"classic" layout, so I went with "better reblogs", fitting the other
"better" extensions.

**Comment by citrusella** [comment]
Maybe a name like "Better Reblog Looks", to try to emphasize that Editable Reblogs are elsewhere and this just takes care of dashboard visuals? Or "Better Dashboard Style" or something? Trying to figure out a name to throw out there that describes the extension well while also not misrepresenting it as a way to unlock reblog editing...

**Comment by ThePsionic** [comment]
My main issue (read: nitpick) about this naming is that "Reblogs", in my mind, refers to the act of reblogging, eg. being in a post window to reblog a post. That's why I went with "Classic Comments".

**Comment by dlmarquis** [comment]
"Comment Options"? To match "Header Options", which was originally named "Classic Header".

**Comment by Wolvan** [comment]
I like @dlmarquis' idea

**Comment by nightpool** [comment]
otoh, most people pretty intuitively connect the "brokenness" to reblogs,
and I don't think they'd have the same intuitive understanding about what a
"comment options" does.

**Comment by Wolvan** [comment]
@nightpool I have seen lots of people call them threads, how about we call them Thread Options instead?

**Comment by blackjackkent** [comment]
Maybe "Better Reblog Display" (or "Post Display" or "Dashboard Design")?

**Comment by macleodsawyer** [comment]
I am in support for **Reblog Options**, because most people (IE 13-17 year olds) on tumblr call threads and comments on tumblr 'reblogs'. 

It just makes sense to go with the name that majority know it as.

**Comment by blackjackkent** [comment]
Reblog Display Options, maybe? (I think the "display" part is critical for distinguishing from ER).

**Comment by ThePsionic** [comment]
:+1: :point_up:

**Comment by Wolvan** [comment]
That does sound good to me :+1:

**Comment by rimbaum** [comment]
Really digging Reblog Display Options. Enough to distinguish itself from ER, and intuitive enough to let people know what it does.

**Comment by macleodsawyer** [comment]
👍

**Comment by nightpool** [comment]
Addressed name stuff and other outstanding issues. I don't think there are any more nitpicks?

please leave line comments if so—there's too much stuff in the discussion to be able to follow it well.

**Comment by nightpool** [comment]
if anyone reviews this tonight, don't merge it. I'll r= it with your name tomorrow morning, so I can write the announcement and be on call when it gets deployed.

File: Extensions/better_reblogs.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
+//* TITLE Reblog Display Options **//
+//* VERSION 1.1 **//
```
Comment (Wolvan) [inline Extensions/better_reblogs.js]: Might want to fix the version to semver format, also why 1.1 and not 1.0?

File: Extensions/better_reblogs.js [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+
+            var all_quotes = [];
+            reblog_tree.find(".reblog-list-item").each(function() {
+                var $this = $(this);
+                var content = $this.find('.reblog-content');
+                var author = $this.find('.reblog-tumblelog-name');
+                var reblog_data = {
+                    reblog_content: content.html() || '',
+                    reblog_author: author.contents()[0].data || '',
+                    reblog_url: author.href() || ''
```
Comment (Wolvan) [inline Extensions/better_reblogs.js]: When I try to switch to or enabled the extension with nested style it tells me
image
.attr("href") confirmed to work

File: Extensions/better_reblogs.js (hunk 1 above)
Comment (nightpool) [inline Extensions/better_reblogs.js]: done. 1.1 was just carried over from another file.

File: Extensions/better_reblogs.js (hunk 2 above)
Comment (nightpool) [inline Extensions/better_reblogs.js]: done.

File: Extensions/better_reblogs.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            XKit.tools.add_css(".reblog-list-item .reblog-title {margin-left:0!important;}", "better_reblogs");
+        }
+
+        XKit.extensions.better_reblogs.do_flat();
+        XKit.post_listener.add("better_reblogs", XKit.extensions.better_reblogs.do_flat);
+
+    },
+
+    run_nested: function() {
+        var match_only_main_dash = "ol" ? !XKit.page.peepr : "";
```
Comment (Wolvan) [inline Extensions/better_reblogs.js]: I think this should be ``var match_only_main_dash = !XKit.page.peepr ? "ol" : "";``

**Comment by blackjackkent** [comment]
This looks good to me except for one minor issue -- when I have it set to nested and switch it BACK to flat, the nested style doesn't disappear:

reblogs

Reloading the page cleans it up, obviously, but still not ideal.

Other than that, LGTM.

**Comment by macleodsawyer** [comment]
Yeah that's been an issue for awhile. Keeps popping back up. 

Didn't someone mention they had a fix earlier?

**Comment by nightpool** [comment]
@macleodsawyer yeah there's basically a fix but I realized that its doubled or tripled in complexity because of how tumblr serializes off-screen posts.

considering just forcing a refresh after changing the main "style" pref tbh

**Comment by bcsiefert** [comment]
I have a simple fix:
inside of the do_flat function in the $(posts).each
```javascript
            var $this = $(this);
            //This is what was added
            if ($this.find("div.old_reblogs.post_body")){
                $("div.old_reblogs.post_body").remove();            
            }
            //end of what was added
```
I tested it on chrome, but it should work on every browser, unless jquery has specific browser-only features that I don't know about...

**Comment by nightpool** [comment]
That won't work for posts cached by Tumblr using Ghost just, as I mentioned
in my previous comment.

Try scrolling down a few pages, running that, and scrolling to the top
again.

**Comment by dlmarquis** [comment]
I think we should just release this at this point, and mention somewhere that it requires a refresh. It's not a big enough issue to keep this in PR purgatory.

**Comment by Wolvan** [comment]
I agree with @dlmarquis, @nightpool would that be alright?

**Comment by nightpool** [comment]
Yeah that sounds great. I'll have it auto refresh the page when you change
the layout type, with a warning.

**Comment by dlmarquis** [comment]
@nightpool I don't want to be an ass, but this really needs to be out. If you can't get to this by tomorrow night (Wednesday), I think I'll just r+ this and add the changes in another pull.

**Comment by dlmarquis** [comment]
Okay, @homu r+

**Pull request merged**